Telemetry: Add physical topology API for Ethernet links

kkfernandezTT opened the pull request:
### Ticket
Solves #31528

  ### Problem description
  Telemetry lacked visibility into the physical topology of Ethernet links. Users couldn't determine:
  - Whether a link is internal (same host) or external (cross-host)
  - Which physical port/connector a link uses on the board
  - The datacenter location (aisle, rack) of remote endpoints

  This information is critical for debugging connectivity issues, understanding cluster topology, and mapping logical metrics to physical infrastructure.

  ### What's changed
  Added a comprehensive physical link topology API to telemetry:

  **Core Data Structures:**
  - `EthernetEndpoint`: Uniquely identifies endpoints by (tray, ASIC location, channel), excluding `chip_id` which changes
  per run
  - `RemoteEndpointInfo`: Captures remote endpoint details including hostname, logical coordinates, and datacenter location (aisle, rack)
  - `PhysicalLinkInfo`: Encapsulates physical port information with a factory method and derived `is_external()` query that enforces the invariant (`remote_endpoint` present iff external)

  **`TopologyHelper` Integration:**
  - Extended `TopologyHelper` to parse FSD connections and build endpoint-to-physical-port mappings
  - Maps ASIC channels to physical ports using Board API
  - Validates endpoints and handles errors gracefully
  - Caches Board objects for efficiency

  **Telemetry Integration:**
  - Added Ethernet metric path validation with proper prefix checking (`tray{n}/chip{m}/channel{l}`)
  - Populates `physical_link_info` in initial snapshots (immutable data, sent once per metric)
  - Supports merging across distributed snapshots with conflict detection
  - Backward compatible JSON serialization

  **Modern C++ Refactoring:**
  - Introduced template helpers (`process_metrics_to_snapshot`, `merge_map_into`) to eliminate code duplication across metric types (`bool`, `unsigned`, `double`, `string`)
  - Optimized string path building with pre-allocation
  - Applied `std::string_view`, `std::optional`, and structured bindings throughout
  - Reduced ~244 lines while adding functionality

  **Testing:**
  - Added 50+ unit tests covering data structures, hashing, parsing, validation, merging, and API behavior
  - Tests use mocks to avoid UMD/FSD dependencies
  - Comprehensive edge case and error path coverage

  **Impact:** Users can now query physical topology for any Ethernet metric, enabling better debugging, infrastructure
  mapping, and operational visibility.


### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

btrzynadlowski-tt commented on tt_telemetry/include/topology/topology.hpp:
```diff
@@ -11,33 +11,116 @@
+struct PhysicalLinkInfo {
+    // Physical port on this board
+    tt::scaleout_tools::PortType port_type{};
+    tt::scaleout_tools::PortId port_id{};
```
Question: is port_id the physical port number on the machine that a human would use for identificaton?

kkfernandezTT committed:
Add physical port and topology info to Ethernet telemetry

nlohmann::json serializes std::unordered_map<uint16_t, std::string> as arrays,
not objects. Update test fixtures to use the correct format.

kkfernandezTT committed:
Refactor Ethernet metrics to use immutable label design

kkfernandezTT committed:
Address review: restore hierarchical paths, remove EthernetMetricBase, remove accidental file

btrzynadlowski-tt commented on tt_telemetry/telemetry/ethernet/ethernet_metrics.cpp:
```diff
@@ -264,18 +336,24 @@ EthernetUncorrectedCodewordCountMetric::EthernetUncorrectedCodewordCountMetric(
 
 const std::vector<std::string> EthernetUncorrectedCodewordCountMetric::telemetry_path() const {
-    return endpoint_telemetry_path(tray_id_, asic_location_, channel_, "uncorrectedCodewordCount");
+    return {
```
Why did this get inlined and endpoint_telemetry_path() removed?

kkfernandezTT committed:
Add helper function for Ethernet endpoint telemetry paths

Extract repeated path construction into build_ethernet_endpoint_path() helper
to reduce duplication across all 5 Ethernet metric types.
